### PR TITLE
Add CA1862 and CA1863

### LINKED ISSUE #38266: Rule "CA1863 Cache a 'CompositeFormat' for repeated use in this formatting operation" is missing a documentation page
### Type of issue

Missing information

### Description

After migrating a project from .Net 6 to .Net 8, I started to see the new code quality rule "CA1863". I wanted to see how to fix it and what is the reason behind this new rule.

The shown warning : 
image

I clicked on the code to open the documentation page and it showed me an error (404 - Page not found). 

The given URL : https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1863

The given error : 
image

I suspect that the page wasn't writing yet because .Net 8 just launched.

### Page URL

https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules

### Content source URL

https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1863.md

### Document Version Independent Id

507d7a47-c1b0-f708-f1fe-1088927110f6

### Article author

gewarren

### PR DESCRIPTION
**Pull request by gewarren**
Contributes to #34171
Fixes https://github.com/dotnet/docs/issues/38266

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/fundamentals/code-analysis/quality-rules/ca1862.md](https://github.com/dotnet/docs/blob/b337052d53910e52417fa5fa3fcedbb5f3c98708/docs/fundamentals/code-analysis/quality-rules/ca1862.md) | [CA1862: Use the 'StringComparison' method overloads to perform case-insensitive string comparisons](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1862?branch=pr-en-us-38344) |
| [docs/fundamentals/code-analysis/quality-rules/ca1863.md](https://github.com/dotnet/docs/blob/b337052d53910e52417fa5fa3fcedbb5f3c98708/docs/fundamentals/code-analysis/quality-rules/ca1863.md) | ["CA1863: Use 'CompositeFormat'"](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1863?branch=pr-en-us-38344) |
| [docs/fundamentals/code-analysis/quality-rules/index.md](https://github.com/dotnet/docs/blob/b337052d53910e52417fa5fa3fcedbb5f3c98708/docs/fundamentals/code-analysis/quality-rules/index.md) | [Code quality rules](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/index?branch=pr-en-us-38344) |
| [docs/fundamentals/code-analysis/quality-rules/performance-warnings.md](https://github.com/dotnet/docs/blob/b337052d53910e52417fa5fa3fcedbb5f3c98708/docs/fundamentals/code-analysis/quality-rules/performance-warnings.md) | [Performance rules](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/performance-warnings?branch=pr-en-us-38344) |

### CODE REVIEW

File: docs/fundamentals/code-analysis/quality-rules/ca1863.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+The following example shows two violations of the rule:
+
+```csharp
+class C
+{
+    private static readonly string StaticField;
+
+    static void Main()
+    {
+        _ = string.Format(StaticField, 42);
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1863.md:47]: @stephentoub I was unable to get the rule to fire on the equivalent VB code. Here's what I tried:

```vb
Class C
    Private Shared ReadOnly s1 As String
    Sub M1()

        Dim unused = String.Format(s1, 42)

        Dim sb As New StringBuilder()
        sb.AppendFormat(s1, 42)
    End Sub
End Class
```

**Review by BillWagner — APPROVED** [review]
This LGTM @gewarren 

Let's :shipit:

File: docs/fundamentals/code-analysis/quality-rules/ca1862.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+Code compares two strings in a case-insensitive manner by first calling <xref:System.String.ToLower>, <xref:System.String.ToLowerInvariant>, <xref:System.String.ToUpper>, or <xref:System.String.ToUpperInvariant> on one or both strings.
+
+## Rule description
+
+When code calls <xref:System.String.ToLower>, <xref:System.String.ToLowerInvariant>, <xref:System.String.ToUpper>, or <xref:System.String.ToUpperInvariant>, an allocation is performed. If the only reason for calling these methods is to perform a case-insensitive string comparison or search, the allocation is unnecessary. Instead, you can call a string comparison method that takes a <xref:System.StringComparison> and specify one of the `*IgnoreCase` values.
+
+## How to fix violations
+
+Remove the call to <xref:System.String.ToLower>, <xref:System.String.ToLowerInvariant>, <xref:System.String.ToUpper>, or <xref:System.String.ToUpperInvariant>, and either call one of the <xref:System.StringComparer> methods, or one of the following methods that takes a <xref:System.StringComparison> argument:
+
```
Comment (carlossanlop) [inline docs/fundamentals/code-analysis/quality-rules/ca1862.md:35]: The CA1862 behavior change was that we won't offer a fix when the user combines cultures (for example: ToUpper vs ToUpperInvariant).

We also updated the diagnostic description to include a warning for the user telling them that they need to test the fix if they decide to apply it, or use Ordinal if needed.

File: docs/fundamentals/code-analysis/quality-rules/ca1863.md (hunk 1 above)
Comment (stephentoub) [reply docs/fundamentals/code-analysis/quality-rules/ca1863.md:47]: It's quite possible it doesn't work for VB currently. It was only tested for C#.

**Pull request merged**